PipelineRun with pipelineSpecs panics

# Expected Behavior

An error or a working state

# Actual Behavior

A crash : 

<details>
  <summary>Crash log</summary>
<pre>
{"level":"info","logger":"tekton","caller":"metrics/prometheus_exporter.go:37","msg":"Created Opencensus Prometheus exporter with config: &{tekton.dev/pipeline tekton prometheus 5000000000 9090  false false  }. Start the server for Prometheus exporter."}
{"level":"info","logger":"tekton","caller":"metrics/config.go:243","msg":"Successfully updated the metrics exporter; old config: <nil>; new config &{tekton.dev/pipeline tekton prometheus 5000000000 9090  false false  }"}
{"level":"info","logger":"tekton","caller":"sharedmain/main.go:150","msg":"Starting informers."}
{"level":"info","logger":"tekton","caller":"sharedmain/main.go:156","msg":"Starting controllers..."}
{"level":"info","logger":"tekton.pipeline-controller","caller":"controller/controller.go:280","msg":"Starting controller and workers","knative.dev/controller":"pipeline-controller"}
{"level":"info","logger":"tekton.pipeline-controller","caller":"controller/controller.go:290","msg":"Started workers","knative.dev/controller":"pipeline-controller"}
{"level":"info","logger":"tekton.pipeline-controller","caller":"pipelinerun/pipelinerun.go:114","msg":"Reconciling 2019-12-23 12:22:38.271615838 +0000 UTC m=+20.559734881","knative.dev/controller":"pipeline-controller"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1470ca3]

goroutine 370 [running]:
github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.(*Reconciler).reconcile(0xc00074e370, 0x1bc3840, 0xc000544c30, 0xc000c44fc0, 0x16, 0xc000c69f60)
	/go/src/github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:297 +0x963
github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.(*Reconciler).Reconcile(0xc00074e370, 0x1bc3840, 0xc000544c30, 0xc000c883c0, 0x2c, 0xc000c70340, 0x1bc3840)
	/go/src/github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:179 +0x67b
knative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc0007cb180, 0xc00032c700)
	/go/src/github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:335 +0x680
knative.dev/pkg/controller.(*Impl).Run.func1(0xc000c4d0a0, 0xc0007cb180)
	/go/src/github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:285 +0x5b
created by knative.dev/pkg/controller.(*Impl).Run
	/go/src/github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:283 +0x1ef
</pre>
</details>

# Steps to Reproduce the Problem

1. Import buildah task
2. Create PipelineResource for `git` and `image` output
3. Create a PipelineRun with pipelineSpec like this : 

```yaml
---
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: pipeline-uno
spec:
  resources:
  - name: git
    resourceRef:
      name: repo-ein
  - name: image
    resourceRef:
      name: image-dois

  pipelineSpec:
    tasks:
      - name: build-ehad
        taskRef:
          name: buildah
```

4. Observe controller output

# Additional Info

This is using last night build from git, available [here](https://raw.githubusercontent.com/openshift/tektoncd-pipeline/release-next/openshift/release/tektoncd-pipeline-nightly.yaml) as well

--- a/pkg/reconciler/pipelinerun/pipelinerun.go
+++ b/pkg/reconciler/pipelinerun/pipelinerun.go
@@ -283,7 +283,7 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 			Status: corev1.ConditionFalse,
 			Reason: ReasonFailedValidation,
 			Message: fmt.Sprintf("Pipeline %s can't be Run; it has an invalid spec: %s",
-				fmt.Sprintf("%s/%s", pipelineMeta.Namespace, pr.Name), err),
+				fmt.Sprintf("%s/%s", pipelineMeta.Namespace, pipelineMeta.Name), err),
 		})
 		return nil
 	}
@@ -295,7 +295,7 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 			Status: corev1.ConditionFalse,
 			Reason: ReasonInvalidBindings,
 			Message: fmt.Sprintf("PipelineRun %s doesn't bind Pipeline %s's PipelineResources correctly: %s",
-				fmt.Sprintf("%s/%s", pr.Namespace, pr.Name), fmt.Sprintf("%s/%s", pr.Namespace, pr.Spec.PipelineRef.Name), err),
+				fmt.Sprintf("%s/%s", pr.Namespace, pr.Name), fmt.Sprintf("%s/%s", pr.Namespace, pipelineMeta.Name), err),
 		})
 		return nil
 	}
@@ -322,7 +322,7 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 			Status: corev1.ConditionFalse,
 			Reason: ReasonParameterTypeMismatch,
 			Message: fmt.Sprintf("PipelineRun %s parameters have mismatching types with Pipeline %s's parameters: %s",
-				fmt.Sprintf("%s/%s", pr.Namespace, pr.Name), fmt.Sprintf("%s/%s", pr.Namespace, pr.Spec.PipelineRef.Name), err),
+				fmt.Sprintf("%s/%s", pr.Namespace, pr.Name), fmt.Sprintf("%s/%s", pr.Namespace, pipelineMeta.Name), err),
 		})
 		return nil
 	}

--- a/pkg/reconciler/pipelinerun/pipelinerun_test.go
+++ b/pkg/reconciler/pipelinerun/pipelinerun_test.go
@@ -291,6 +291,9 @@ func TestReconcile_InvalidPipelineRuns(t *testing.T) {
 			tb.TaskInputs(tb.InputsParamSpec("some-param", v1alpha1.ParamTypeString)))),
 		tb.Task("a-task-that-needs-array-params", "foo", tb.TaskSpec(
 			tb.TaskInputs(tb.InputsParamSpec("some-param", v1alpha1.ParamTypeArray)))),
+		tb.Task("a-task-that-needs-a-resource", "ns", tb.TaskSpec(
+			tb.TaskInputs(tb.InputsResource("workspace", "git")),
+		)),
 	}
 	ps := []*v1alpha1.Pipeline{
 		tb.Pipeline("pipeline-missing-tasks", "foo", tb.PipelineSpec(
@@ -320,6 +323,18 @@ func TestReconcile_InvalidPipelineRuns(t *testing.T) {
 		tb.PipelineRun("pipeline-resources-not-declared", "foo", tb.PipelineRunSpec("a-pipeline-that-should-be-caught-by-admission-control")),
 		tb.PipelineRun("pipeline-mismatching-param-type", "foo", tb.PipelineRunSpec("a-pipeline-with-array-params", tb.PipelineRunParam("some-param", "stringval"))),
 		tb.PipelineRun("pipeline-conditions-missing", "foo", tb.PipelineRunSpec("a-pipeline-with-missing-conditions")),
+		tb.PipelineRun("embedded-pipeline-resources-not-bound", "foo", tb.PipelineRunSpec("", tb.PipelineRunPipelineSpec(
+			tb.PipelineTask("some-task", "a-task-that-needs-a-resource"),
+			tb.PipelineDeclaredResource("workspace", "git"),
+		))),
+		tb.PipelineRun("embedded-pipeline-invalid", "foo", tb.PipelineRunSpec("", tb.PipelineRunPipelineSpec(
+			tb.PipelineTask("bad-t@$k", "b@d-t@$k"),
+		))),
+		tb.PipelineRun("embedded-pipeline-mismatching-param-type", "foo", tb.PipelineRunSpec("", tb.PipelineRunPipelineSpec(
+			tb.PipelineParamSpec("some-param", v1alpha1.ParamTypeArray),
+			tb.PipelineTask("some-task", "a-task-that-needs-array-params")),
+			tb.PipelineRunParam("some-param", "stringval"),
+		)),
 	}
 	d := test.Data{
 		Tasks:        ts,
@@ -365,6 +380,18 @@ func TestReconcile_InvalidPipelineRuns(t *testing.T) {
 			name:        "invalid-pipeline-missing-conditions-shd-stop-reconciling",
 			pipelineRun: prs[7],
 			reason:      ReasonCouldntGetCondition,
+		}, {
+			name:        "invalid-embedded-pipeline-resources-bot-bound-shd-stop-reconciling",
+			pipelineRun: prs[8],
+			reason:      ReasonInvalidBindings,
+		}, {
+			name:        "invalid-embedded-pipeline-bad-name-shd-stop-reconciling",
+			pipelineRun: prs[9],
+			reason:      ReasonFailedValidation,
+		}, {
+			name:        "invalid-embedded-pipeline-mismatching-parameter-types",
+			pipelineRun: prs[10],
+			reason:      ReasonParameterTypeMismatch,
 		},
 	}
 
@@ -394,7 +421,12 @@ func TestReconcile_InvalidPipelineRuns(t *testing.T) {
 				t.Errorf("Expected failure to be because of reason %q but was %s", tc.reason, condition.Reason)
 			}
 			if !tc.hasNoDefaultLabels {
-				expectedLabels := map[string]string{pipeline.GroupName + pipeline.PipelineLabelKey: tc.pipelineRun.Spec.PipelineRef.Name}
+				expectedPipelineLabel := tc.pipelineRun.Name
+				// Embedded pipelines use the pipelinerun name
+				if tc.pipelineRun.Spec.PipelineRef != nil {
+					expectedPipelineLabel = tc.pipelineRun.Spec.PipelineRef.Name
+				}
+				expectedLabels := map[string]string{pipeline.GroupName + pipeline.PipelineLabelKey: expectedPipelineLabel}
 				if len(tc.pipelineRun.ObjectMeta.Labels) != len(expectedLabels) {
 					t.Errorf("Expected labels : %v, got %v", expectedLabels, tc.pipelineRun.ObjectMeta.Labels)
 				}

--- a/test/builder/pipeline.go
+++ b/test/builder/pipeline.go
@@ -386,24 +386,37 @@ func PipelineRunNilTimeout(prs *v1alpha1.PipelineRunSpec) {
 	prs.Timeout = nil
 }
 
-// PipelineRunNodeSelector sets the Node selector to the PipelineSpec.
+// PipelineRunNodeSelector sets the Node selector to the PipelineRunSpec.
 func PipelineRunNodeSelector(values map[string]string) PipelineRunSpecOp {
 	return func(prs *v1alpha1.PipelineRunSpec) {
 		prs.PodTemplate.NodeSelector = values
 	}
 }
 
-// PipelineRunTolerations sets the Node selector to the PipelineSpec.
+// PipelineRunTolerations sets the Node selector to the PipelineRunSpec.
 func PipelineRunTolerations(values []corev1.Toleration) PipelineRunSpecOp {
 	return func(prs *v1alpha1.PipelineRunSpec) {
 		prs.PodTemplate.Tolerations = values
 	}
 }
 
-// PipelineRunAffinity sets the affinity to the PipelineSpec.
+// PipelineRunAffinity sets the affinity to the PipelineRunSpec.
 func PipelineRunAffinity(affinity *corev1.Affinity) PipelineRunSpecOp {
 	return func(prs *v1alpha1.PipelineRunSpec) {
 		prs.PodTemplate.Affinity = affinity
+	}
+}
+
+// PipelineRunPipelineSpec adds a PipelineSpec to the PipelineRunSpec.
+// Any number of PipelineSpec modifiers can be passed to transform it.
+func PipelineRunPipelineSpec(ops ...PipelineSpecOp) PipelineRunSpecOp {
+	return func(prs *v1alpha1.PipelineRunSpec) {
+		ps := &v1alpha1.PipelineSpec{}
+		prs.PipelineRef = nil
+		for _, op := range ops {
+			op(ps)
+		}
+		prs.PipelineSpec = ps
 	}
 }
 

--- a/test/builder/pipeline_test.go
+++ b/test/builder/pipeline_test.go
@@ -264,6 +264,35 @@ func TestPipelineRunWithResourceSpec(t *testing.T) {
 	}
 }
 
+func TestPipelineRunWithPipelineSpec(t *testing.T) {
+	pipelineRun := tb.PipelineRun("pear", "foo", tb.PipelineRunSpec("", tb.PipelineRunPipelineSpec(
+		tb.PipelineTask("a-task", "some-task")),
+		tb.PipelineRunServiceAccountName("sa"),
+	))
+
+	expectedPipelineRun := &v1alpha1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pear",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.PipelineRunSpec{
+			PipelineRef: nil,
+			PipelineSpec: &v1alpha1.PipelineSpec{
+				Tasks: []v1alpha1.PipelineTask{{
+					Name:    "a-task",
+					TaskRef: v1alpha1.TaskRef{Name: "some-task"},
+				}},
+			},
+			ServiceAccountName: "sa",
+			Timeout:            &metav1.Duration{Duration: 1 * time.Hour},
+		},
+	}
+
+	if diff := cmp.Diff(expectedPipelineRun, pipelineRun); diff != "" {
+		t.Fatalf("PipelineRun diff -want, +got: %s", diff)
+	}
+}
+
 func TestPipelineResource(t *testing.T) {
 	pipelineResource := tb.PipelineResource("git-resource", "foo", tb.PipelineResourceSpec(
 		v1alpha1.PipelineResourceTypeGit, tb.PipelineResourceSpecParam("URL", "https://foo.git"),

